rosbag2_transport/src/rosbag2_transport/player.cpp

```diff
@@ -1191,17 +1187,61 @@ bool PlayerImpl::publish_message(rosbag2_storage::SerializedBagMessageSharedPtr
     return message_published;
   }
 
+  // Try to publish message as service request
   auto client_iter = service_clients_.find(message->topic_name);
   if (client_iter != service_clients_.end()) {
-    if (!client_iter->second->is_include_request_message(*message->serialized_data)) {
-      return message_published;
+    const auto & service_client = client_iter->second;
+    // TODO(morlov):
+    //  Wrap deserialize_service_event and get_service_event_type_and_client_gid in try-catch
+    auto service_event = service_client->deserialize_service_event(*message->serialized_data);
+    if (!service_event) {
+      RCLCPP_ERROR_STREAM(
+        owner_->get_logger(), "Failed to deserialize service event message for '" <<
+          service_client->get_service_name() << "' service!\n");
+      return false;
+    }
+
+    auto [service_event_type, client_gid] =
+      service_client->get_service_event_type_and_client_gid(service_event);
+    // Ignore response message
+    if (service_event_type == service_msgs::msg::ServiceEventInfo::RESPONSE_SENT ||
+      service_event_type == service_msgs::msg::ServiceEventInfo::RESPONSE_RECEIVED)
+    {
+      // TODO(morlov): Shall we ignore REQUEST_RECEIVED as well?
+      return false;
+    }
+
+    if (!service_client->generic_client()->service_is_ready()) {
+      RCLCPP_ERROR(
+        owner_->get_logger(), "Service request hasn't been sent. The '%s' service isn't ready !",
+        service_client->get_service_name().c_str());
+      return false;
+    }
+
+    if (!service_client->is_service_event_include_request_message(service_event)) {
```

Comment:
It appears to overlook the scenario where both the service and the client enable introspection contents for the same client ID.

Old function is PlayerServiceClient::is_include_request_message().  
Except checking if request is included, this function also includes the following logic
1.  Decide whether the request comes from the service or the client is used for one client ID. (Currently, it is decided by the first message including the request data. e.g. First message with request from client.  We will use message from client to send request, even if message from service also provide request data.) 

2. Consider the case. Service/client switched introspection type from metadata to contents in recorded data. 
I don't consider the case. Service/Client switched introspection type from contents to metadata in record data, which leads that the decision in 1 maybe adjusted.

https://github.com/Barry-Xu-2018/rosbag2/blob/a8312de491a34150ddcb0be4156b7e01ba2ac862/rosbag2_transport/src/rosbag2_transport/player_service_client.cpp#L55-L159

Reply:
@Barry-Xu-2018 Thanks for finding and pointing out this. It seems I overlook these cases.
From one side I understand that it would be nice to have some autodetection algorithm to be able to decide whether to use a service request from the original client request or from the service request confirmation event message. 
However, this sophisticated logic could have corner cases and would be difficult to understand or explain for final users.
In addition, it seems the original implementation with a "hidden" autodetection algorithm from the `PlayerServiceClient::is_include_request_message()` is not going to work due to the found issue on the RMW layer [[service introspection event does not have unique GID during transaction](https://github.com/ros2/rmw/issues/357#top)
#357](https://github.com/ros2/rmw/issues/357).

I am more inclined to implement a simple logic for a while to allow publish service requests upon a boolean flag by client request or by service confirmation and add this flag to the settings and perhaps to the CLI or env variable. Need to figure out what is better on the current stage.

Also, I think that we need to publish original service events in all other cases to be able to see them in the `topic echo` during playback or other introspection tools.

cc: @fujitatomoya

Reply:
@MichaelOrlov 

> However, this sophisticated logic could have corner cases and would be difficult to understand or explain for final users.

Yes. Since users are not aware of the internal selection logic, it can lead to confusion about the result under certain scenario.

> I am more inclined to implement a simple logic for a while to allow publish service requests upon a boolean flag by client request or by service confirmation and add this flag to the settings and perhaps to the CLI or env variable. Need to figure out what is better on the current stage.

I also agree with using simpler logic. However, this might result in losing support for complex scenarios, such as some services enabling introspection while others have client introspection enabled. It would require users to uniformly use either service introspection or client introspection.  

I incline to add new option for play.   
e.g.
--service-request-from {service, client}. `service` is default value. 
If there's a need to add automatic selection in the future (This depends on whether there is a demand from the users.), we can add an "auto" option.   

> Also, I think that we need to publish original service events in all other cases to be able to see them in the topic echo during playback or other introspection tools.

Are you saying that users, besides being able to replay service requests, can also replay the service event topic like a normal topic, allowing other introspection tools to analyze recorded the information via service event topic?

If yes, enable this function, is there a new parameter for play (if this parameter exists, play service event topic instead of play service request) ?  About filter parameters (--services, --exclude-services, etc), do they work? Or user should use --topics, --exclude-topics since service event topic is replayed as normal topic.)

Reply:
>  About filter parameters (--services, --exclude-services, etc), do they work? Or user should use --topics, --exclude-topics since service event topic is replayed as normal topic.)

After checking current code, `--services` and `--exclude-services` should still function normally, affecting the playback of service events.

Reply:
I try to add a new parameter `{service_introspection,client_introspection}` for play command.   
For the details, refer to https://github.com/Barry-Xu-2018/rosbag2/commit/4d15a21e8e67c7a8dbe2265a3ff5ab1671171ef8

Reply:
I try to add a new parameter `--publish-service-event` to support playing service event message.  
For the deails, refer to https://github.com/Barry-Xu-2018/rosbag2/commit/006bfa9fbc33f23b277f986e3935afdbe5764a36

Reply:
@Barry-Xu-2018 Thanks for the follow-up and deduction on my proposals. I am sorry I was off for a couple of days. Now, I'm trying to recover.  
As regards adding a new parameter `--service-request-from {service_introspection,client_introspection}` for the play command.
- Yes, the idea is good, and the implementation seems overall correct. I might need to make some minor changes, like nitpicks or review it thoroughly. I will cherry-pick it to my WIP branch and add my edits above as separate commits.

As regards adding a new parameter `--publish-service-event` to support playing service event messages.  
IMO it will be more consistent to publish services as service events by default since we are saving them this way (just as is) and have an option to `--publish-service-requests`. In this case our other newly added option `--service-request-from {service_introspection,client_introspection}` will be applicable only if the `--publish-service-requests` is enabled.

Reply:
@MichaelOrlov 

> Yes, the idea is good, and the implementation seems overall correct. I might need to make some minor changes, like nitpicks or review it thoroughly. I will cherry-pick it to my WIP branch and add my edits above as separate commits.

Okay

> IMO it will be more consistent to publish services as service events by default since we are saving them this way (just as is) and have an option to --publish-service-requests. In this case our other newly added option --service-request-from {service_introspection,client_introspection} will be applicable only if the --publish-service-requests is enabled.

Okay.  I start to update https://github.com/Barry-Xu-2018/rosbag2/commit/006bfa9fbc33f23b277f986e3935afdbe5764a36

Reply:
@MichaelOrlov 

I updated code https://github.com/Barry-Xu-2018/rosbag2/commit/d6db1bb7cb026d99f41061947e3296d0055a7a5a to support `--publish-service-requests`

Reply:
@Barry-Xu-2018 Sorry didn't have time today to review or cherry-pick your latest updates from https://github.com/Barry-Xu-2018/rosbag2/commit/d6db1bb7cb026d99f41061947e3296d0055a7a5a to support `--publish-service-requests`.

I added API to be able to make tests more deterministic [Add Player::wait_for_sent_service_requests_to_finish() API](https://github.com/Barry-Xu-2018/rosbag2/pull/6/commits/9d98b04e725c7a5e987e930848b9d75d0817d81a).
I would appreciate it if you would rewrite other service-related unit tests in a similar way as I made in the [Make tests play_service_requests_from_service(client) deterministic](https://github.com/Barry-Xu-2018/rosbag2/pull/6/commits/b87c944d5933079c2d61213275b915384d0094b2)

Reply:
@MichaelOrlov 

Thanks. 

> I would appreciate it if you would rewrite other service-related unit tests in a similar way as I made in the [Make tests play_service_requests_from_service(client) deterministic](https://github.com/Barry-Xu-2018/rosbag2/pull/6/commits/b87c944d5933079c2d61213275b915384d0094b2)

Okay. I will update other service-related unit tests by this way.

Reply:
@Barry-Xu-2018 Ok I've cherry-picked (https://github.com/Barry-Xu-2018/rosbag2/commit/d6db1bb7cb026d99f41061947e3296d0055a7a5a to support `--publish-service-requests`) and made some fixes above it. See my new commits on this branch.

Reply:
@MichaelOrlov 

> I will update other service-related unit tests by this way.

I updated test code based on latest code. Please refer to https://github.com/Barry-Xu-2018/rosbag2/commit/e09632321a0cc4433a193e4c54559e6da6ff0657.

Reply:
@Barry-Xu-2018 I've cherry-picked your changes in tests from https://github.com/Barry-Xu-2018/rosbag2/commit/e09632321a0cc4433a193e4c54559e6da6ff0657. Honestly, didn't have time to review thoroughly, but this shouldn't be a problem. 
If you don't have any more comments/requests for changes for This PR let's move forward and merge rhi PR to your branch. 
I think we almost good to go. Will need to rebase and run CI after merge.

Reply:
I have some nitpick in mind to fix - but we can do it after code freeze